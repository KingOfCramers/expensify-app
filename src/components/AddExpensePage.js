import React from 'react';
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startAddExpense } from "../actions/expenses";

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense); // Passing this in as a prop (using the mapdispatch to props ) to allow for easier testing, rather than using global dispatch function.
    this.props.history.push("/")
  }
  render(){
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h2 className="page-header__title">Add Expense</h2>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  }
};

const mapDispatchToProps = (dispatch) => { // We have access to dispatch thanks to react-redux. The goal is to simplify testing, by
  return {
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
  }
};
export default connect(undefined, mapDispatchToProps)(AddExpensePage); // Second argument is matchDispatchToProps