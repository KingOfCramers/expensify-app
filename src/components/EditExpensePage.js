import React from 'react';
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, startRemoveExpense } from "../actions/expenses";

/// GIT LEARNING TO ROLLBACK
export class EditExpensePage extends React.Component {

  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push("/")
  };

  onRemove = () => {
    this.props.startRemoveExpense({id: this.props.expense.id }) // .then(() => { console.log("This worked!")} ) // You could chain on an action here!
    this.props.history.push("/")
  };
 //
  render(){
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button onClick={this.onRemove}>Remove</button>
      </div>
    );
  }
};

const mapStateToProps = (state, props) => { // we have access to the props as the second argument here.
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id ) // the props.match.params comes from the Router (it's part of the URL)
  }
};

const mapDispatchToProps = (dispatch, props) => {
  return {
    startRemoveExpense: (expense) => { return dispatch(startRemoveExpense(expense))}, // This returns a promise (the bottomn line is implicit return)... See the promise return example above.
    editExpense: (id, expense) => dispatch(editExpense(id, expense))
  }
}


export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);