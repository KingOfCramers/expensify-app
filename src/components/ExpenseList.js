import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";
// Unconnected //
export const ExpenseList = (props) => ( // Exporting for testing...
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-body">
      {props.expenses.length === 0 ? (
        <div className="list-item--empty">
          <p>No expenses</p>
        </div>
      ) : (
        props.expenses.map((expense) => {
          return <ExpenseListItem key={expense.id} {...expense}/>
        })
      )}
    </div>
  </div>
);

const mapStateToProps = (state) => { // This lets us determine what aspects of the redux state we want to pass in.
  return {
    expenses: selectExpenses(state.expenses, state.filters) // Use the values on our state object to calculate the visible expenses.
  }
}

export default connect(mapStateToProps)(ExpenseList); // "connect" is a higher order component.