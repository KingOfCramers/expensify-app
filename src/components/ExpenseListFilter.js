import { connect } from "react-redux";
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from "../actions/filters";
import { DateRangePicker } from "react-dates";
import React from "react";

export class ExpenseListFilter extends React.Component {
  state = {
    calendarFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }))
  };

  onSortChange = (e) => {
    if(e.target.value === "date"){
      this.props.sortByDate();
    } else {
      this.props.sortByAmount();
    }
  };

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value)
  };

  render(){
    return (
      <div>
          <input type="text" value={this.props.filters.text} onChange={this.onTextChange} />
          <select onChange={this.onSortChange}>
            <option value="date">Date</option>
            <option value="amount">Amount</option>
          </select>
          <DateRangePicker
            startDate={this.props.filters.startDate} // Accepts moment object
            startDateId={"fluffykins"}
            endDateId={"fluffykins"}
            endDate={this.props.filters.endDate} // Accepts moment object
            onDatesChange={this.onDatesChange} // passes startDate and endDate to function, to set new state.
            focusedInput={this.state.calendarFocused}
            onFocusChange={this.onFocusChange} // Sets focus state to calendarFocused state.
            showClearDates={true}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
        </div>
    );
  }
}

const mapStateToProps = (state) => { // This lets us determine what aspects of the redux state we want to pass in.
  return {
    filters: state.filters
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate))
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);