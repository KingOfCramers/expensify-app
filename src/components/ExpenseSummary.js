import React from "react";
import { connect } from "react-redux";
import selectExpenses from "../selectors/expenses";
import getExpensesTotal from "../selectors/expensesTotal";
import numeral from "numeral";
import { Link } from "react-router-dom";

export const ExpenseSummary = ({ count, total }) => { // Export for testing purposes
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">You're viewing <span>{count}</span> expenses with a value of <span>{numeral(total/100).format("$0,0.00")}</span>.</h1>
        <div className="page-header__actions">
          <Link to="/create" className="button">Add Expense</Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    count: selectExpenses(state.expenses, state.filters).length,
    total: getExpensesTotal(selectExpenses(state.expenses, state.filters)),
  }
};

export default connect(mapStateToProps)(ExpenseSummary);