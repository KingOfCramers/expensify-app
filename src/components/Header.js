import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from "react-redux";
import { startLogout } from "../actions/auth"; // A function...

export const Header = ({ startLogout }) => ( // Props need to be imported in a stateless functional here!
  <header>
    <h1>Expensify</h1>
    <NavLink exact to="/" activeClassName="is-active">Home</NavLink>
    <NavLink to="/create" activeClassName="is-active">Create</NavLink>
    <NavLink to="/help" activeClassName="is-active">Help</NavLink>
    <button onClick={startLogout}>Log out</button>
  </header>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);