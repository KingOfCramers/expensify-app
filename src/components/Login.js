import React from "react";

class LoginPage extends React.Component {
  state =  {
    email: "",
    password: ""
  }

  onEmailChange() {
    console.log("Loaded")
  }

  onPasswordChange(){
    console.log("Loaded")
  }

  render(){
    return (
      <div>
        <form>
          <input
            type="text"
            placeholder="Email"
            autoFocus
            value={this.state.email}
            onChange={this.onEmailChange}
          />
          <input
            type="text"
            placeholder="Password"
            value={this.state.password}
            onChange={this.onPasswordChange}
          />
          <button>Login</button>
        </form>
      </div>
    );
  };
};

export default LoginPage;