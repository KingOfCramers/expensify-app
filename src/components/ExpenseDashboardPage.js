import React from 'react';
import ExpenseList from "./ExpenseList";
import ExpenseListFilter from "./ExpenseListFilter";
import ExpenseSummary from "./ExpenseSummary";
import { connect } from "react-redux";

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseSummary />
    <ExpenseListFilter />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;