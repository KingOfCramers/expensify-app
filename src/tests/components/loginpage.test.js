import React from "react";
import { shallow } from "enzyme";
import { LoginPage } from "../../components/LoginPage"; // Make sure to get named exports to avoid problems w/ connect.
import thunk from 'redux-thunk';

test("Should render login page", () => {
  const wrapper = shallow(<LoginPage />);
  expect(wrapper).toMatchSnapshot();
});

test("Should call startLogin on button click", () => {
  const login = jest.fn();
  const wrapper = shallow(<LoginPage startLogin={login}/>);
  wrapper.find("button").simulate("click");
  expect(login).toHaveBeenCalled();
});