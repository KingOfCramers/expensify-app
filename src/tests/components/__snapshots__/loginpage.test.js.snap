// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render login page 1`] = `
<div>
  <form>
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Password"
      type="text"
      value=""
    />
    <button>
      Login
    </button>
  </form>
</div>
`;
